stream-chat-android-sample/src/main/kotlin/io/getstream/chat/sample/feature/channel/ChannelFragment.kt

```diff
@@ -18,26 +19,41 @@ import kotlinx.android.synthetic.main.fragment_channel.*
 
 class ChannelFragment : Fragment(R.layout.fragment_channel) {
 
+    private val cid: String by lazy { navArgs<ChannelFragmentArgs>().value.cid }
+
+    private val viewModelFactory: ChannelViewModelsFactory by lazy { ChannelViewModelsFactory(cid) }
+
+    private val messagesViewModel: MessageListViewModel by viewModels { viewModelFactory }
+
+    private val channelHeaderViewModel: ChannelHeaderViewModel by viewModels { viewModelFactory }
+
+    private val messageInputViewModel: MessageInputViewModel by viewModels { viewModelFactory }
+
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
-        val cid = navArgs<ChannelFragmentArgs>().value.cid
+        initMessagesViewModel()
+        initHeaderViewModel()
+        initMessageInputViewModel()
 
-        val messagesViewModel = MessageListViewModel(cid)
-            .apply { bindView(messageListView, viewLifecycleOwner) }
-            .apply {
-                state.observe(
-                    viewLifecycleOwner,
-                    Observer {
-                        when (it) {
-                            is MessageListViewModel.State.Loading -> showProgressBar()
-                            is MessageListViewModel.State.Result -> hideProgressBar()
-                            is MessageListViewModel.State.NavigateUp -> navigateUp()
-                        }
+        val backButtonHandler = {
+            messagesViewModel.onEvent(MessageListViewModel.Event.BackButtonPressed)
+        }
+
+        channelHeaderView.onBackClick = { backButtonHandler() }
+
+        activity?.apply {
+            onBackPressedDispatcher.addCallback(
```

Comment:
Never seen this before. Very cool!

Reply:
New to me, too. I just extracted what was already there to a method.